pkg/diagnostics/tracing.go

```diff
@@ -264,6 +270,83 @@ func getAPIComponent(apiPath string) apiComponent {
 	return apiComponent{componentType: tokens[1], componentValue: tokens[2]}
 }
 
+// GetSpanAttributesMapFromGRPC builds the span trace attributes map for gRPC calls based on given parameters as per open-telemetry specs
+func GetSpanAttributesMapFromGRPC(req interface{}, rpcMethod string) map[string]string {
+	// RPC Span Attribute reference https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md
+	// gRPC method /package.service/method
+	var serviceName string
+
+	p := rpcMethod
+	if p[0] == '/' {
+		p = rpcMethod[1:]
+	}
+
+	// Split up to 3 delimiters in '/package.service/method'
+	// example for internal call : "/dapr.proto.internals.v1.ServiceInvocation/CallLocal"
+	// example for non-internal call : "/dapr.proto.runtime.v1.Dapr/InvokeService"
+	internalCall := strings.Contains(p, "dapr.proto.internals.")
+
+	tokens := strings.SplitN(p, "/", 3)
+	if len(tokens) >= 2 {
+		if internalCall {
+			if t := strings.SplitN(tokens[0], ".", 5); len(t) >= 4 {
+				serviceName = t[4]
+			}
+		} else {
+			// TODO : need to revisit /package.service/method format in runtime calls, currently service name is "Dapr"
+			// so instead of taking general "Dapr", taking method name as service name to give better insights
+			serviceName = tokens[1]
+		}
+	} else {
+		// default service name to full method name, after removing first "/"
+		serviceName = p
+	}
+
+	m := make(map[string]string)
+	m[gRPCServiceSpanAttributeKey] = serviceName
+
+	// below attribute is not as per open temetery spec, adding custom dapr attribute to have additional details
+	m[gRPCDaprInstanceSpanAttributeKey] = extractComponentValueFromGRPCRequest(req)
+	return m
+}
+
+func extractComponentValueFromGRPCRequest(req interface{}) string {
+	if req == nil {
```

Comment:
We can simplify all `if`s with switch.

```
switch req.(type) {
case internalv1pb.InternalInvokeRequest:
  return s.Message.GetMethod()
case runtimev1pb.PublishEventRequest:
  return s.GetTopic()
...
}
```

Reply:
@youngbupark , I want to use type assertion syntax to avoid runtime panic, which can not be used outside switch apparently. So I need to perform type assertion in each case, to get the member accessed , which is making more verbose, so adopted this way. Please let me know if still it can be done that I am not aware  of or missing something.

Reply:
type switch-case doesn't panic. We can use type assertion in each case. type assertion in each cases will not make difference because each `if` is doing type assertion. 

IMO, there are two reasons that I suggest to use switch-case:

1. Readability - switch-cases is more readable than multiple `if`s. 
2. Performance - switch-case can avoid multiple type-assertions. if req is the type `runtimev1pb.BindingEventRequest`, then the code will do meaningless nine type assertions until it reaches to the last if. swich-case can avoid this situation. As far as I understand, type conversion or assertion in go creates new object. if req size is big with `runtimev1pb.BindingEventRequest` type, then it could impact the performance because it will do all ten type conversions.

Reply:
ok, it seems type switch case internally do type assertions. Also I no need to use type assertion variant with in each case as it is already checked on switch case, I can simply convert to desired object. I changed the code now. Thanks !!

Reply:
fixed the lint issue, found that we can assign in the type switch and can actually avoid each switch statement casting..this is great.